pkg/doc/jwt/jwt.go

```diff
@@ -0,0 +1,459 @@
+}
+
+// Signed creates JWS.
+func (b *builder) Signed(signer Signer) (JSONWebToken, error) {
+	if b.err != nil {
+		return nil, b.err
+	}
+
+	// build headers
+	headers := signer.Headers()
+
+	err := checkJWSHeaders(headers)
+	if err != nil {
+		return nil, fmt.Errorf("check JOSE headers: %w", err)
+	}
+
+	sigInput, err := signingInput(headers, b.claims)
+	if err != nil {
+		return nil, fmt.Errorf("prepare JWT verification data: %w", err)
+	}
+
+	signature, err := signer.Sign(sigInput)
+	if err != nil {
+		return nil, fmt.Errorf("sign JWT verification data: %w", err)
+	}
+
+	return &jsonWebToken{
+		headers:   headers,
+		payload:   b.claims,
+		signature: signature,
+	}, nil
+}
+
+func checkJWSHeaders(headers map[string]interface{}) error {
+	if _, ok := headers[HeaderAlgorithm]; !ok {
+		return fmt.Errorf("%s headers is not defined", HeaderAlgorithm)
+	}
+
+	if _, ok := headers[HeaderType]; !ok {
+		headers[HeaderType] = TypeJWT
+	}
+
+	return nil
+}
+
+// Unsecured creates unsecured JWT.
+func (b *builder) Unsecured(headers map[string]interface{}) (JSONWebToken, error) {
+	if b.err != nil {
+		return nil, b.err
+	}
+
+	return &jsonWebToken{
+		headers: PrepareJWTUnsecuredHeaders(headers),
+		payload: b.claims,
+	}, nil
+}
+
+// New initiates creation of a new JSON Web Token. claims could be anything marshallable to the map.
+func New(claims interface{}) Builder {
+	m, err := toMap(claims)
+
+	return &builder{
+		claims: m,
+		err:    err,
+	}
+}
+
+// FromSigned parses JWS in compact serialization format.
+// verifier can be nil, in this case signature verification is not made.
+func FromSigned(jws string, verifier Verifier, detachedPayload []byte) (JSONWebToken, error) {
+	parts := strings.Split(jws, ".")
+	if len(parts) != jwtPartsCount {
+		return nil, errors.New("invalid JWT compact format")
+	}
+
+	headers, err := base64.RawURLEncoding.DecodeString(parts[jwtHeaderPart])
+	if err != nil {
+		return nil, fmt.Errorf("decode base64 header: %w", err)
+	}
+
+	var jwtHeaders map[string]interface{}
+
+	err = json.Unmarshal(headers, &jwtHeaders)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal JSON headers: %w", err)
+	}
+
+	var payload []byte
+	if len(detachedPayload) == 0 {
+		payload, err = base64.RawURLEncoding.DecodeString(parts[jwtPayloadPart])
+		if err != nil {
+			return nil, fmt.Errorf("decode base64 payload: %w", err)
+		}
+	} else {
+		payload = detachedPayload
+	}
+
+	var jwtPayload map[string]interface{}
+
+	err = json.Unmarshal(payload, &jwtPayload)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal JSON payload: %w", err)
+	}
+
+	signature, err := base64.RawURLEncoding.DecodeString(parts[jwtSignaturePart])
+	if err != nil {
+		return nil, fmt.Errorf("decode base64 signature: %w", err)
+	}
+
+	jwt := &jsonWebToken{
+		headers:   jwtHeaders,
+		payload:   jwtPayload,
+		signature: signature,
+	}
+
+	if verifier != nil {
+		err = verifier.Verify(jwt)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return jwt, nil
+}
```

Comment:
nit: better to avoid nested conditions
```
if verifier == nil {
 return jwt, nil
}

if err = verifier.Verify(jwt); err != nil {
    return nil, err
}

return jwt, nil
```

Reply:
agree